rfcs/2023-12-19-ADMIN_AUTH.md

```diff
@@ -0,0 +1,302 @@
+# Admin Authentication Rev.2
+
+Feature Name: Admin Authentication
+
+Proposal Date: 2023-12-20
+
+Author: Hugh Francis
+
+## Description
+
+**Note:** See below for Rev.1 to understand the full decision matrix that led to this revision.
+
+Rev.2 describes an approach to authentication that uses what we're preliminarily calling the
+"Mothership API". This is a central API that dome.fm instances register their presence with,
+and periodically backup their state to.
+
+Another advantage of the dome.fm Mothership API is that it will house SMTP credentials, so 
+that dome.fm instances do not need to BYO their own SMTP (albeit in the future this will be 
+configurable in the case the adminstrator would prefer that soverignty).
+
+## Why Next.js for Mothership API?
+
+In a new codebase `github.com/dome.fm/mothership`, we'll setup a Next.js app. 
+
+Why Next.js for an API? Well, for a few tactical reasons:
+- This app will likely live at `developers.dome.fm`, and we may want to render HTML there (think
+API documentation).
+- Dome.fm will be a Next.js app, so this team will already have that proficiency.
+- The Mothership API needs to be highly available, so serverless Vercel deployment is a great
+option. Next.js & Vercel play super nice.
+
+## How Dome.fm instances will register themselves
+
+1. In a [Next.js Middleware](https://nextjs.org/docs/app/building-your-application/routing/middleware), 
+we'll ask the Dome.fm instance if it's registered itself (we'll likely check the PostgresQL 
+database for a `MothershipAPIKey`, ie `const hasMothershipAPIKey = !!(await prisma.mothershipAPIKey.findMany())[0]`)
+
+2. If `false`, we'll `POST https://developers.dome.fm/instance-api/[...domeInstanceVersion]/instances` with a body
+like:
+
+```
+{
+  owner: "hughfrancis89@gmail.com"
+}
+```
+
+3. The Mothership API will return a response like:
+
+```
+{
+  instanceId: "some-uuid-generated-by-mothership",
+  mothershipAPIKey: "some-secret-key-generated-by-mothership"
+}
+```
+
+4. The Dome.fm instance will save the returned `MothershipAPIKey` in PostgresQL and continue into 
+request stack.
+
+## How Dome.fm instances will trigger Auth Emails
+
+**Note:** Rather than providing an open SMTP server for the instance to authenticate and
+connect to, we'll offer an API that triggers a pre-formatted authentication email (rather 
+than an open SMTP server which could be more widely abused).
```

Comment:
Would this flow be any different in the future, when an artist could supply their own SMTP creds? Seems like a safe assumption to me...I guess somebody might want to override the rate limit in this case, but can't imagine why 5/hr wouldn't be sufficient.

Reply:
Ya, if an artist supplies their own SMTP creds, we'd essentially "undock" the instance, and their auth email flow would just go direct to their instance.

& Ya, a rate limiter wouldn't be quite as important in that internal case because it's all internal requests (not a public API)